### PR TITLE
v3.8 Support openharmony4.0

### PR DESCRIPTION
**Pull request by qiuguohua**
Re: #

### Changelog

Deveco3.1.0.501 supports openharmony 3.x, supports openharmony4.x with harmonyos.

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Review by PatriceJiang — APPROVED** [review]

File: templates/openharmony/entry/build-profile.json5 [hunk 1]
```diff
@@ -10,11 +10,12 @@
...
       "cppFlags": "",
     },
-    "compileMode": "esmodule",
+    "compileMode": "jsbundle",
   },
   "targets": [
     {
-      "name": "default"
+      "name": "default",
+      "runtimeOS": "HarmonyOS"
```
Comment (qiuguohua) [inline templates/openharmony/entry/build-profile.json5:18]: Must use the harmonyos runtime environment.

File: platforms/native/builtin/index.js [hunk 2]
```diff
@@ -211,6 +211,12 @@ for (const key in jsbWindow) {
     }
 }
 
+// In the openharmony platform, XMLHttpRequest is not undefined, but there are problems to using it.
+// So the native implementation is forced to be used.
+if (window.oh && globalThis.XMLHttpRequest != undefined) {
```
Comment (PPpro) [inline platforms/native/builtin/index.js]:
```suggestion
if (window.oh && typeof globalThis.XMLHttpRequest !== 'undefined') {
```

Comment (PPpro) [reply platforms/native/builtin/index.js]: does it mean XMLHttpRequest is supported on openharmony 

but there are some bugs on their XMLHttpRequest implementation ?

File: templates/openharmony/entry/build-profile.json5 (hunk 1 above)
Comment (PPpro) [inline templates/openharmony/entry/build-profile.json5:18]: what about the OS is OpenHarmony?

Comment (qiuguohua) [reply templates/openharmony/entry/build-profile.json5:18]: I think this should be a bug that is not available to environments using Openharmony. We can only use harmonyos.

File: platforms/native/builtin/index.js (hunk 2 above)
Comment (qiuguohua) [reply platforms/native/builtin/index.js]: The Openharmony team says it doesn't provide an XMLHttpRequest interface, but XMLHttpRequest isn't undefined. there are problems with using it.

Comment (PPpro) [reply platforms/native/builtin/index.js]: that means we keep two different implementation of `XMLHttpRequest`, it's strange

where the  globalThis.XMLHttpRequest comes from ?

File: platforms/native/builtin/index.js [hunk 3]
```diff
@@ -211,6 +211,12 @@ for (const key in jsbWindow) {
     }
 }
 
+// In the openharmony platform, XMLHttpRequest is not undefined, but there are problems to using it.
+// So the native implementation is forced to be used.
+if (window.oh && typeof globalThis.XMLHttpRequest !== 'undefined') {
+    globalThis.XMLHttpRequest = jsbWindow.XMLHttpRequest;
+}
```
Comment (minggo) [inline platforms/native/builtin/index.js:218]: How other native platforms set `globalThis.XMLHttpRequest`?

File: templates/openharmony/entry/build-profile.json5 (hunk 1 above)
Comment (minggo) [reply templates/openharmony/entry/build-profile.json5:18]: It seems wired.

**Review by minggo — APPROVED** [review]

File: platforms/native/builtin/index.js (hunk 2 above)
Comment (minggo) [reply platforms/native/builtin/index.js]: Can create an issue for it.

Comment (PPpro) [reply platforms/native/builtin/index.js]: please setup an issue to figure out where the globalThis.XMLHttpRequest is defined

**Review by PPpro — APPROVED** [review]

**Pull request merged**